### ISSUE TITLE
Destroy and re-create table yields misaligned headers.

### ISSUE THREAD

**Issue by kushal-likhi** [handsontable/handsontable#2748]
Hi,

If i create a Handsontable with a set of data it renders fine. 
Then when we destroy table with `.destroy()` method and re-create a table by running exact same function as used originally with same data,  we get a table rendered with misaligned column header.

If page is refreshed all works good. But on second render after destroy it breaks.

Below is the settings and screenshots:

## Settings
Below are the settings. All references values are actually functions, if you need their code let me know.
```javascript
this.plugins.handsonTable._renderedTable = new Handsontable(this.config.tableGridContainer.$el.get(0), {
                    data: this._parsedFileData,
                    minSpareRows: 0,
                    fixedColumnsLeft: 0,
                    invalidCellClassNameString: this.config.table.invalidCellClassName,
                    colHeaders: this.plugins.handsonTable.generators.colHeaders,
                    rowHeaders: this.plugins.handsonTable.generators.rowHeaders,
                    contextMenu: false,
                    height: this.config.tableGridContainer.$el.height(),
                    width: this.config.tableGridContainer.$el.width(),
                    afterRenderer: this.plugins.handsonTable.renderer.afterCellRenderDecorator,
                    beforeRender: this.plugins.handsonTable.renderer.beforeRender,
                    afterRender: this.plugins.handsonTable.renderer.afterRender,
                    cells: this.plugins.handsonTable.renderer.cells,
                    manualColumnResize: true,
                    manualRowResize: true,
                    autoRowSize: true,
                    autoColumnSize: true,
                    multiSelect: true,
                    wordWrap: false,
                    afterChange: this.plugins.handsonTable.eventHandlers.afterChange,
                    afterSelection: this.plugins.handsonTable.eventHandlers.afterSelection,
                    maxRows: this._parsedFileData.length,
                    maxCols: this._parsedFileData[0].length
                });
```

#### First Render (All good):

![screen shot 2015-08-22 at 2 18 19 pm](https://cloud.githubusercontent.com/assets/571089/9423213/be68f918-48d8-11e5-8c9b-42a05b4a92cc.png)

#### Second Render (Misaligned):

![screen shot 2015-08-22 at 2 17 54 pm](https://cloud.githubusercontent.com/assets/571089/9423216/d628a53a-48d8-11e5-99b5-c8384ae27cc7.png)

What could be the reason? any hints?

**Comment by kushal-likhi** [comment]
Handsontable version: `0.16.1`

**Comment by AMBudnik** [comment]
Hi @kushal-likhi it would be great if you could create a working demo for this example. Thanks in advance.

**Comment by kushal-likhi** [comment]
Hi, @AMBudnik thanks for a quick reply. Here is a jsfiddle describing the problem.  https://jsfiddle.net/h4774en5/

Thanks for the help.

**Comment by kushal-likhi** [comment]
Could it be due to some `Walkontable` caches?? Or any hints of the cause?

**Comment by AMBudnik** [comment]
So at the beggining we have

![first](https://cloud.githubusercontent.com/assets/10757813/9443638/7081457e-4a82-11e5-9182-db580547b7e4.png)

and after recreating we get

![second](https://cloud.githubusercontent.com/assets/10757813/9443645/79f96f50-4a82-11e5-88b4-f032749e6750.png)

There is something also changing with a fourth's row header.

We will take a closer look at this issue. Please stay tunned.

**Comment by kushal-likhi** [comment]
Thanks a lot. :) meanwhile some hints or places for me to look at, such that i could be of some help?

**Comment by kushal-likhi** [comment]
**A hint:** This happens when we have `fixedColumnsLeft: < 1`. Values `0` or `not supplied` causes this issue. If we have the value set to `1 or more` this issue is not there.

**Comment by kushal-likhi** [comment]
**Another Find**: Has to do with caches, or some variables not cleared on destroy. 

If i refresh the script via loading it *after destroy and before re-render*, then it works.

Using below function to reload (For testing):
```javascript
//Src is handsontable.js file link.
function refreshScript (src) {
  var scriptElement = document.createElement('script');
  scriptElement.type = 'text/javascript';
  scriptElement.src = src;
  document.getElementsByTagName('head')[0].appendChild(scriptElement);
}
```

Hope this helps.

**Comment by kushal-likhi** [comment]
Thanks guys.. :)

**Comment by gillius** [comment]
I'm having this issue or an extremely similar one (my table looks like the screenshots posted by @AMBudnik. I see this in Firefox always but haven't in Chrome yet and it behaves differently depending on autoRowSize true or false and whether or not I am pasting many rows vs. adding them. I see the rows are 22.5px tall but the row headers are 22px tall.

I see this has a committed fix, and I'm very happy to see that because this might be the only issue left for our app to adopt HOT. Is 0.17 expected out within a month? If not I will try to backport to 0.16.1 or attempt to find a "stable" version of 0.17 branch to deploy.

**Comment by budnix** [comment]
Hi @gillius this patch will be released tomorrow as 0.17.0 version.

**Comment by gillius** [comment]
Wow, @bundix, very good then! I will eagerly test it as soon as I see it come out.

**Comment by kushal-likhi** [comment]
@budnix Awesome, thanks.. really appreciate.

**Comment by krzysztofspilka** [comment]
Works fine now.

**Comment by gillius** [comment]
Perhaps I have a different bug, but 0.17.0 doesn't work for me; in fact may be worse now. I updated via bower and confirmed in Firefox debug console that HOT header is 0.17.0. In the attached images I show the top and bottom of the table. I can reproduce this easily via copy and paste. I can see that the headers are shifted so much that the first row disappears, and at the bottom I have two copies of row headers (and the table extends outside of its containing div, overlapping the next block below. Sometimes editing the other cells in the table can make the row headers get less shifted, but still not shifted. This happens whether autoRowSize is true or false, but seems to be even worse with false. With 0.16.1 it would shift half a pixel per row but with 0.17.0 the entire top row header disappears and also still has 0.5px/row offset.

If this is a different issue and I should file a separate issue, let me know (I'm not actually ever destroying, recreating, or calling updateSettings on this table).

![tabletop](https://cloud.githubusercontent.com/assets/1524332/9520291/ed911c70-4c93-11e5-8dc9-5dea370a09fd.PNG)
...
![tablebottom](https://cloud.githubusercontent.com/assets/1524332/9520290/ed8c0ee2-4c93-11e5-8076-cb81d43909af.PNG)

**Comment by budnix** [comment]
@gillius Can you prepare demo with this bug on jsfiddle?

**Comment by gillius** [comment]
I did some more investigation and found two separate issues:

1. I'm using Bootstrap 3 and it sets "margin: 4px 0px 0px" on input[type="checkbox"]. If I have a checkbox in my row in Firefox, it adds 0.5px height to the row but NOT to the row header. If I disable the margin, it works.
2. I had issues getting minSpareRows to work how I wanted, so in my afterChange I do a data.push to add a "spare" row in certain conditions, then call HOT instance render method. This causes the row headers to shift. For whatever reason it's lot worse and always breaks in 0.17.0 when pasting. However, I will assume that this is my fault and there is a better way for me to implement spare rows so I'll investigate workarounds.

For the first item, this happens using a fiddle based on the "basic" example from the HOT site by adding a checkbox column (http://jsfiddle.net/rcc0k0d2/). In this case, the http://handsontable.com/styles/main.css sets input[type="checkbox"] margin to "0px 0px 1rem", which for whatever reason in Firefox 40.0.2 ends up giving a td height of 32.5, 0.5px smaller than the 33px row header. This works in Chrome 44. I don't understand why a margin at an integer px would cause a 0.5px shift in td height in Firefox.

If you'd like, I can create a separate issue for the "0.5px" row height difference or the data manipulation in afterChange (if that's not a user error). In my case, the workaround is simple and I will apply a class to remove Bootstrap's checkbox margin within the scope of my HOT.

**Comment by gillius** [comment]
1. I fixed the issue of adding rows in afterChange by calling hot.alter('insert_row') instead of pushing a new row into the bound data model and calling render. I'm still not sure if that is supposed to be kosher but it works perfectly.
2. The CSS fix to remove the margin on checkboxes fixes the Firefox "0.5px" issue. I'm still willing to file an issue if you think HOT should be able to handle fractional px row heights.
